### PR TITLE
Add permission + rejection logic to Consumer

### PR DESCRIPTION
**Pull request by eelcovdw**
## Description

Changes in consumer:
- SyncDecision + SyncDecisionTuple are combined to SyncAction to bundle execute + reject logic per action type
- action validation logic now throws an error instead of returning (bool, str)
- updated consumer to match above changes

Sync rejection:
- happens when action.execute() receives a 403 from server
- each action implements its rejection behavior
- rejections are logged as REJECTED in local_state

TODO:
- Integration tests for rejection logic (once serverside permissions are live)
- Rejected files are only retried when the local/remote file changes (and not when the permission changes)
- better place for initial download_batch?

## Affected Dependencies
List any dependencies that are required for this change.

## How has this been tested?
- Describe the tests that you ran to verify your changes.
- Provide instructions so we can reproduce.
- List any relevant details for your test configuration.

## Checklist
- [ ] I have followed the [Contribution Guidelines](https://github.com/OpenMined/.github/blob/master/CONTRIBUTING.md) and [Code of Conduct](https://github.com/OpenMined/.github/blob/master/CODE_OF_CONDUCT.md)
- [ ] I have commented my code following the [OpenMined Styleguide](https://github.com/OpenMined/.github/blob/master/STYLEGUIDE.md)
- [ ] I have labeled this PR with the relevant [Type labels](https://github.com/OpenMined/.github/labels?q=Type%3A)
- [ ] My changes are covered by tests

### CODE REVIEW

File: syftbox/client/plugins/sync/local_state.py [hunk 1]
```diff
@@ -34,6 +35,28 @@ class LocalState(BaseModel):
...
+    def insert_completed_action(self, action: SyncAction) -> None:
+        """Insert action result into local state."""
+        if action.action_type == SyncActionType.NOOP:
+            return
+
+        if action.status == SyncStatus.PROCESSING:
+            raise ValueError("Attempted to insert an action into LocalState that is still being processed.")
+
+        if action.status == SyncStatus.SYNCED:
+            self.insert_synced_file(
```
Comment (koenvanderveen) [inline syftbox/client/plugins/sync/local_state.py:47]: just out of interest, what is the difference here?

File: syftbox/client/plugins/sync/sync_action.py [hunk 2]
```diff
@@ -0,0 +1,356 @@
...
+    action_type = SyncActionType.CREATE_LOCAL
+
+    def execute(self, client: SyncClient) -> None:
+        content_bytes = client.download(self.path)
+        abs_path = client.workspace.datasites / self.path
+        abs_path.parent.mkdir(parents=True, exist_ok=True)
+        abs_path.write_bytes(content_bytes)
+        self.status = SyncStatus.SYNCED
+
+    def reject(self, client: SyncClient, reason: Optional[str] = None) -> None:
```
Comment (koenvanderveen) [inline syftbox/client/plugins/sync/sync_action.py:174]: I think this is more like `process_rejection`?

File: syftbox/client/plugins/sync/sync_action.py [hunk 3]
```diff
@@ -0,0 +1,356 @@
...
+            # TODO error handling
+            raise ValueError("Hash mismatch after applying diff")
+
+        # TODO implement safe write with tempfile + rename
+        abs_path.parent.mkdir(parents=True, exist_ok=True)
+        abs_path.write_bytes(new_data)
+        self.status = SyncStatus.SYNCED
+
+    def reject(self, client: SyncClient, reason: Optional[str] = None) -> None:
+        # Client doesnt have read permission, so the local file is deleted
```
Comment (koenvanderveen) [inline syftbox/client/plugins/sync/sync_action.py:202]: why do we have to delete here?

File: syftbox/client/plugins/sync/sync_action.py [hunk 4]
```diff
@@ -0,0 +1,356 @@
...
+    if (
+        action.action_type in {SyncActionType.CREATE_REMOTE, SyncActionType.MODIFY_REMOTE}
+        and action.local_metadata.file_size > max_size_bytes
+    ):
+        raise SyncValidationError(f"File {abs_path} is larger than {MAX_FILE_SIZE_MB}MB.")
+
+    # Create/modify with broken permissions is invalid
+    if (
+        action.action_type in {SyncActionType.CREATE_REMOTE, SyncActionType.MODIFY_REMOTE}
+        and SyftPermission.is_permission_file(abs_path)
```
Comment (koenvanderveen) [inline syftbox/client/plugins/sync/sync_action.py:337]: we need to refactor this soon but thats oke

**Review by koenvanderveen — APPROVED** [review]

File: syftbox/client/plugins/sync/local_state.py (hunk 1 above)
Comment (eelcovdw) [reply syftbox/client/plugins/sync/local_state.py:47]: Just moved this to a different place, and added a check so we dont insert PROCESSING accidentally

**Pull request merged**